### PR TITLE
SYCL TeamPolicy scratch memory alternative

### PR DESCRIPTION
**Pull request by masterleinad**
Based on #3759. Corresponding to #3694. It seems #3759 is the way to go but I want to get some confirmation before closing #3694. I would also prefer to keep this separate from #3759 and merge that one first.

### CODE REVIEW

**Review by nliber — APPROVED** [review]

File: core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp:389]: Why are you doing `std::max(shmem_size, 0)`?

File: core/unit_test/incremental/Test12a_ThreadScratch.hpp
Comment (dalg24) [inline core/unit_test/incremental/Test12a_ThreadScratch.hpp:122]: Is that the only way to disable test 12a?

Comment (masterleinad) [reply core/unit_test/incremental/Test12a_ThreadScratch.hpp:122]: I think so. 
The above comment is also not true with this version. I just wanted to confirm that I get the same functionality as in the pull request corresponding to the current TeamPolicy implementation. 
IMHO, it's fine without fixing it here but I'm definitely working on it.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp:389]: Safeguarding against `shmen_size` being negative by any chance, e.g. when no scratch memory is requested, resulting in a tremendous amount of memory being allocated.

Comment (dalg24) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp:389]: > e.g. when no scratch memory is requested

Then it would be set to zero
https://github.com/kokkos/kokkos/blob/5a0b850ca52a60722e5cd5efe65264be72f50ac6/core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp#L233

This does not feel like the right place for check invariants of your class.  Also that probably something you should not be doing silently.

Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp]: I simply removed it.

File: core/unit_test/incremental/Test12a_ThreadScratch.hpp
Comment (masterleinad) [reply core/unit_test/incremental/Test12a_ThreadScratch.hpp:122]: I updated the comment.

File: core/unit_test/incremental/Test12b_TeamScratch.hpp
Comment (rgayatri23) [inline core/unit_test/incremental/Test12b_TeamScratch.hpp:112]: This might lead to merge conflicts with #3748 . Both the scratch related incremental tests are being updated there too with respect to their problem sizes. So one of us might have to wait until the other one gets merged.

Comment (masterleinad) [reply core/unit_test/incremental/Test12b_TeamScratch.hpp:112]: I am used to having merge conflicts after almost every of my pull requests being merged. :slightly_smiling_face: No big deal for me to rebase and fix the conflicts if necessary.

**Comment by masterleinad** [comment]
Retest this please.

**Review by dalg24 — CHANGES_REQUESTED** [review]
You need to discuss your design choices.  You allocate both L0 and L1 in local memory (I am using SYCL names) when I believe CUDA has L0 in local memory and L1 in global memory.

File: core/src/SYCL/Kokkos_SYCL_Instance.cpp
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Instance.cpp:118]: I can argue both ways here. Either we try to be consistent with names for corresponding attributes in `Cuda` or `HIP` or we are using `SYCL` specific terms deliberately. I don't have a strong preference.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp:350]: What would help you? More comments?

File: core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp:347]: They also contribute to the local memory I am allocating.

**Comment by masterleinad** [comment]
> You need to discuss your design choices. You allocate both L0 and L1 in local memory (I am using SYCL names) when I believe CUDA has L0 in local memory and L1 in global memory.

We need to revisit this class a couple of times anyway when implementing the remaining features. Hence, my rational was to get something working (and passing the tests) rather quickly and improve in later pull requests. I would rather have confirmed where we want them to be allocated before changing and I believe this can happen in a later pull request.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp
Comment (dalg24) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp:356]: I missed the thread -> team
Please define `team_scratch_size` variable

**Comment by masterleinad** [comment]
Retest this please.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp:410]: Any particular reason to have this here rather than in the constructor and assigning the right address to `m_scratch_ptr`?

File: core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp:363]: I have not spent any effort in exploring if the value returned here is a good choice in general or not. Currently, it isn't even a power of two which I would think is favorable.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp:402]: Done.

File: core/unit_test/incremental/Test12b_TeamScratch.hpp
Comment (masterleinad) [reply core/unit_test/incremental/Test12b_TeamScratch.hpp:122]: Let's see.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp:472]: We need to allocate some memory when we implement `TeamThread`, `ThreadVector` or `TeamVector` reductions, see #3783.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp:410]: Yes, the `sycl::handler` `cgh` is only available in this context.

File: core/unit_test/incremental/Test12b_TeamScratch.hpp
Comment (masterleinad) [reply core/unit_test/incremental/Test12b_TeamScratch.hpp]: Seems to work now.

**Review by nliber — APPROVED** [review]

File: core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp:510]: Please delete the copy constructor and copy assignment.  This is a bug otherwise.

**Review by dalg24 — COMMENTED** [review]
You can go ahead and clean history after you apply my change request (double free bug)
The rest is fine.

**Comment by masterleinad** [comment]
@dalg24 Actually, just squash all the commits while merging.

**Review by dalg24 — APPROVED** [review]

**Pull request merged**
